### ISSUE TITLE
Questions about implementing model parallelism in the inference engine

### ISSUE THREAD

**Issue by hyunwoongko** [microsoft/DeepSpeed#1161]
Hello. I would like to ask about the model parallelism feature in the inference engine. In general, the model parallelism that I can think of is inter-layer model parallelism like GPipe (only partitioning part, not pipelining) and intra-layer model parallelism like Megatron-LM.

https://github.com/microsoft/DeepSpeed/blob/aa16828c267dc7a790a5cc2f43f93a67617a2364/deepspeed/inference/engine.py#L102

But, in the case of the current implementation of inference engine, it seems to broadcast all parameters from device 0 if mpu is not input. But, as far as I know, these operations don't seem parallelizing way that models are sliced.  

https://github.com/microsoft/DeepSpeed/blob/aa16828c267dc7a790a5cc2f43f93a67617a2364/deepspeed/inference/engine.py#L213

In this part of the code, it seems that input is also broadcasting to all devices. So, I wonder why deepspeed is broadcasting in all parts when mpu isn't inputted.

**Comment by hyunwoongko** [comment]
- num_gpus = 2
![image](https://user-images.githubusercontent.com/38183241/122065262-c2fa5280-ce2c-11eb-927b-9a35cdc7a6c7.png)

- num_gpus = 1
![image](https://user-images.githubusercontent.com/38183241/122065368-da394000-ce2c-11eb-929d-40d75e8d5e77.png)

**Comment by hyunwoongko** [comment]
If model parallelism in the form of tensor slicing for GPT-neo is not currently supported, the paragraph below may need to be slightly modified :) If I'm mistaken, please help me to do model parallelization!

![image](https://user-images.githubusercontent.com/38183241/122068124-166da000-ce2f-11eb-8764-b8b3b8bc8653.png)

**Comment by RezaYazdaniAminabadi** [comment]
For reducing the results across different GPUs, we use the all_reduce in the [inference-api](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/ops/transformer/inference/transformer_inference.py#L311). The model-parallel group is also created in the [inference-engine](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/inference/engine.py#L88) and passed to the inference-module.

**Comment by hyunwoongko** [comment]
Thanks for your kind reply. :)

My question is, why doesn't the memory of each GPU area decrease after slicing the model in two? For example, if we slice a 30GB model into two, the amount of memory allocated to at least one device will have to be less than 30GB, even if it is not completely reduced to 15GB.

Thanks, 
Hyunwoong Ko

**Comment by RezaYazdaniAminabadi** [comment]
The problem is that the model gets created on both GPUs by HunggingFace initially and that takes the model's total memory which is in your case 30GB. However, in deepspeed we partition the parameters and send the corresponding parts to each GPU. So, some of the initially allocated memory is just cached on the GPU and never used. I have not spend time to release that memory in HuggingFace way of initializing a model. By the way, I think you can still use those allocated memory, the issue is that nvidia-smi is not showing the amount of free memory in a precise way.

**Comment by RezaYazdaniAminabadi** [comment]
I have used the torch's memory management (https://pytorch.org/docs/stable/cuda.html#memory-management) and it shows the memory reduction when using model-parallel 2:

```
Before deepspeed initialize: 14042529792 bytes allocated
After deepspeed initialize: 7719548928 bytes allocated
```
I think nvidia-smi is just showing all the cached memory!

Could you please print `{torch.cuda.memory_allocated()}, {torch.cuda.memory_cached()}` before and after deepspeed inference intialize on your side to verify it?
Thanks,
Reza

**Comment by hyunwoongko** [comment]
When I call `torch.cuda.empty_cache()`, memory is reduced as your words. 

![image](https://user-images.githubusercontent.com/38183241/122146786-9a0ca880-ce92-11eb-9d31-bfedd1680c25.png)

Thanks !!
Hyunwoong Ko

**Comment by hyunwoongko** [comment]
Thanks for the very kind reply. Can't we just load the model from the CPU? If I deploy a large model like Blender (9B+) or T5 (10B+), if the hugging face model is loaded on the gpu first, memory allocation will fail.

**Comment by hyunwoongko** [comment]
I success CPU to GPT parallel. Thanks !

**Comment by andreamad8** [comment]
@hyunwoongko any insight on how did you do it?

**Comment by hyunwoongko** [comment]
I developed totally new tools for model parallelism.
My tool can parallelize all the huggingface models. I will open this tool asap.

**Comment by andreamad8** [comment]
ok, thanks!

**Comment by RezaYazdaniAminabadi** [comment]
@hyunwoongko, thanks for pushing to solve all these issues. Please let me know when you are finished with opening this as I am also eager to see your approach.

Best,
Reza

**Comment by hyunwoongko** [comment]
@RezaYazdaniAminabadi 

https://github.com/tunib-ai/parallelformers

We present a novel framework for model parallelization 🎉

**Comment by RezaYazdaniAminabadi** [comment]
@hyunwoongko 

Thanks for sharing this. I will look into it and let you know if I have questions.

**Comment by RezaYazdaniAminabadi** [comment]
Hi @hyunwoongko,

Thanks for sharing the great work you did on parallelizing the transformers models. I think we can use a lot of this in deepspeed to parallelize the different models. 
I went through your description [here](https://github.com/huggingface/transformers/issues/12772#issuecomment-882760778). I agree with the limitation that our approach has, mostly due to using the inference-optimized kernels. However, I think on the parallelism part I see you take more or less the same approach as us. 
I see that you are mainly concerned about the way that we move the model to the different devices (in the inference [engine](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/inference/engine.py#L99)). I think solving this is pretty easy by moving this part after module_inject function which takes care of the model-parallel partitioning.
However, I still see the issue of creating the model on CPU through the HuggingFace Transformers [pipeline](https://github.com/microsoft/DeepSpeedExamples/blob/25d73cf73fb3dc66faefa141b7319526555be9fc/inference/huggingface/gpt-neo.py#L19-L21). That is due to the fact the some of the tensors on the pipeline side (which are not part of the model) will be on CPU whereas the computation of the model is happening on the device side and generates tensors on different devices. 
Please note that using the pipeline way of doing inference is giving a pretty easy of way testing different models that users are normally eager to use.
I have made a [PR ](https://github.com/microsoft/DeepSpeed/pull/1244)to fix this issue on the deepspeed-inference engine. So, we can now create the model on the CPU side, partition it based on the number of GPUs, and then move it to the devices. As you can see the changes are pretty minimal to resolve this limitation!

I would really appreciate if we can work together to bring the parallel implementation design on your side to deepspeed and we can merge it with the high-performance kernels for the different models.

Thanks,
Reza

**Comment by RezaYazdaniAminabadi** [comment]
By the way, you will realize that with this PR the memory of the GPUs won't increase (even the cached one) the same as before.

**Comment by hyunwoongko** [comment]
Thanks for the positive reply. @stas00 and I were discussing the integration of parallelformers and transformers [here](https://github.com/huggingface/transformers/issues/12772). And maybe we're thinking of this as part of the DeepSpeed and Transformers integration, and we want to work on this with the DeepSpeed team. For example, I and you implement and commit in DeepSpeed, and I and Stas utilize this in Huggingface Transformers.

Also, as you said, I'm thinking deeply about how to use fused kernel with the mechanism I'm currently implementing on my side. If this works, we can take the speed of fused kernel and the scalability of my implementation at the same time. Anyway, we need to discuss how to collaborate. How would you like to work?

Thanks,
Hyunwoong Ko

**Comment by hyunwoongko** [comment]
In addition, I also want to implement a training feature, what do you think of this? I think many people want to training Transformers models with the Tensor MP method. I hope ultimately all the models in Transformers support 3D parallelization through ZeRO + Pipeline with Tensor MP.

**Comment by hyunwoongko** [comment]
@RezaYazdaniAminabadi 
I look forward to your reply :)

**Comment by hyunwoongko** [comment]
I'm closing this issue because it's too old and I'll discuss it in a new issue.

**Comment by RezaYazdaniAminabadi** [comment]
Hi @hyunwoongko 

Thanks for the great discussion. I am certainly interested. Let me also discuss this internally, and we can go on with the collaboration soon :)

Thanks,
Reza

**Comment by hyunwoongko** [comment]
Let me know when the discussion is over. I'll start working on it right away :)

**Comment by RezaYazdaniAminabadi** [comment]
Hi @hyunwoongko 

Thanks for checking in :)
I wonder if you will be free for a quick offline chat to go through this?

Thanks,
Reza

**Comment by hyunwoongko** [comment]
Sure! Why don't we arrange the meeting via email? 
My one is kevin.ai@tunib.ai.

**Comment by RezaYazdaniAminabadi** [comment]
Just sent an invite through email. Thanks

**Comment by hyunwoongko** [comment]
Could you summarize what you said at the meeting?

**Comment by hyunwoongko** [comment]
And since the issue here is closed, how about discussing it in new issue?

**Comment by RezaYazdaniAminabadi** [comment]
Yes, better to open another issue, I will send the summary in email.

Thanks,
Reza

**Comment by hyunwoongko** [comment]
https://github.com/microsoft/DeepSpeed/issues/1248

I  already opened new issue !
